apps/b2c/pages/api/auth/[...nextauth].ts

```diff
@@ -0,0 +1,155 @@
+        clientSecret: process.env.FACEBOOK_CLIENT_SECRET!,
+      }),
+      CredentialsProvider({
+        name: 'Credentials',
+        credentials: {
+          username: {
+            label: 'Username',
+            type: 'text',
+            placeholder: 'Username',
+          },
+          password: {
+            label: 'Password',
+            type: 'password',
+            placeholder: 'Password',
+          },
+        },
+        authorize: async (credentials) => {
+          const { username, password } = credentials || {}
+
+          const data = await asyncSSRFetch<{ authenticate: User }>({
+            query: GET_USER_WITH_CREDENTIALS,
+            variables: {
+              username,
+              password,
+            },
+          })
+
+          if (!data.authenticate.id) return null
+
+          const { id, name, role, email, image } = data.authenticate
+          return { id, name, role, email, image }
+        },
+      }),
+    ],
+    jwt: {
+      maxAge: MAX_AGE,
+      encode: async ({ token, secret, maxAge }) => {
+        token = {
+          ...token,
+          exp: getExp({ exp: token?.exp, maxAge }),
+        } as JWT
+
+        const encodedToken = jwt.sign(token, secret, { algorithm: 'HS256' })
+        return Promise.resolve(encodedToken)
+      },
+      decode: async ({ token, secret }) => {
+        if (!token) return null as Awaitable<null>
+
+        const decodedToken = jwt.verify(token, secret, {
+          algorithms: ['HS256'],
+        })
+        return decodedToken as Awaitable<JWT>
+      },
+    },
+    callbacks: {
+      jwt: async ({ token, user }) => {
+        if (user) {
+          const role = user.role ? user.role : 'user'
+
+          token = {
+            ...token,
+            'https://hasura.io/jwt/claims': {
+              'x-hasura-allowed-roles': [role],
+              'x-hasura-default-role': role,
+              'x-hasura-role': role,
+              ...(role === 'user'
+                ? { 'x-hasura-user-id': token.sub }
+                : { 'x-hasura-account-id': token.sub }),
+            },
+          }
+        }
+
+        return Promise.resolve(token)
+      },
+      session: async ({ session, token }) => {
+        const encodedToken = jwt.sign(token, process.env.NEXTAUTH_SECRET!, {
+          algorithm: 'HS256',
+        })
+
+        session.user = {
+          id: token.sub,
+          name: token.name,
+          email: token.email,
+          role: token['https://hasura.io/jwt/claims']['x-hasura-role'],
+          image: token.picture,
+        } as User
+        session.token = encodedToken
+
+        return Promise.resolve(session)
+      },
+    },
+    session: {
+      strategy: 'jwt',
+      maxAge: MAX_AGE,
+    },
+    pages: {
+      signIn: '/',
+      signOut: '/',
+      error: '404',
+    },
+  })
+}
+
+interface GetExpProps {
+  exp?: number
+  maxAge?: number
+}
+
+const getExp = ({ exp, maxAge = MAX_AGE }: GetExpProps) => {
```

Comment:
can we give this a clearer name? like `getExpiry`?

Reply:
updated